docs/style/CODING_STYLE_GUIDE.md

```diff
@@ -0,0 +1,162 @@
+
+We use the following auto-formatters on code:
+
+| Language   | Formatter          | Style File                                                                                                  |
+| ---------- | ------------------ | ----------------------------------------------------------------------------------------------------------- |
+| C++        | clang-format       | [.clang-format](https://github.com/project-chip/connectedhomeip/blob/master/.clang-format)                  |
+| ObjectiveC | clang-format       | [.clang-format](https://github.com/project-chip/connectedhomeip/blob/master/.clang-format)                  |
+| java       | google-java-format | N/A                                                                                                         |
+| Python     | pep8, isort        | [.restyled.yaml](https://github.com/project-chip/connectedhomeip/blob/master/.restyled.yaml) (command line) |
+| YAML       | prettier           | None                                                                                                        |
+| JSON       | prettier           | None                                                                                                        |
+| markdown   | prettier           | None                                                                                                        |
+
+All pull requests run formatting checks using these tools before merge is
+allowed. Generated code is not run through restyle.
+
+### C++
+
+#### Use C++ _cstdint_ for Plain Old Data Types
+
+Standard, scalar data types defined in _cstdint_ should be used for basic signed
+and unsigned integer types, especially when size and serialization to
+non-volatile storage or across a network is concerned.
+
+Examples of these are: `uint8_t`, `int8_t`, etc.
+
+#### Avoid `using namespace` Statements in Headers
+
+By doing this, you are effectively forcing every other module that includes the
+header to also be using the namespace. This causes namespace pollution and
+generally defeats the purposes of namespaces. Fully-qualified symbols should be
+used instead.
+
+#### Classes / objects not exposed in a header should be in an anonymous namespace
+
+If a cpp class defines a class or instantiates a static object, it should be
+enclosed in an anonymous namespace.
+
+```
+namespace {
+  // CPP internal defines go here
+}  // namespace
+```
+
+#### Singleton use
+
+The decision to use a singleton class should be considered with care. Do not
+default to using a singleton for ease of writing code.
+
+If the class truly should be a singleton (ex. if it is controlling access to a
+hardware resource)
+
+-   The standard function name for accessing an SDK singleton is GetInstance().
+-   Singleton classes should delete copy and move constructors
+
+#### Avoid Heap-based Resource Allocation and auto-resizing std containers
+
+Heap-based resource allocation should be avoided. This includes any container
```

Comment:
I think that there is an exception from that rule for platform code which runs on platforms with sufficient amount of RAM. It would be nice to emphasize that this mostly applies to common SDK codebase and RAM-constrained platforms.

Reply:
Good call. Updated this section to indicate this applies only to core SDK code, and controller and platforms may allow heap allocation.